agent/util/expansions.go

```diff
@@ -47,3 +61,30 @@ func (e *DynamicExpansions) Get(key string) string {
 
 	return e.Expansions.Get(key)
 }
+
+// Redact marks the expansion with given key for redaction.
+func (e *DynamicExpansions) RedactKey(key string) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if value := e.Expansions.Get(key); value != "" {
+		e.redact = append(e.redact, RedactInfo{Key: key, Value: value})
+	}
+}
+
+// GetRedacted gets the expansions that should be redacted.
+func (e *DynamicExpansions) GetRedacted() []RedactInfo {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+
+	return e.redact
+}
+
+// PutAndRedact puts the expansions followed by marking it for redaction.
+func (e *DynamicExpansions) PutAndRedact(key, value string) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	e.Expansions.Put(key, value)
+	e.redact = append(e.redact, RedactInfo{Key: key, Value: value})
```

Comment:
Since it looks like we checked this for `PutAndRedact`, can we also only redact here if the value is non-empty?